executor/prebackup_utils.go

```diff
@@ -0,0 +1,121 @@
+package executor
+
+import (
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	k8upv1alpha1 "github.com/vshn/k8up/api/v1alpha1"
+	"github.com/vshn/k8up/cfg"
+)
+
+// fetchPreBackupPodTemplates fetches all PreBackupPods from the same namespace as the originating backup.
+func (b *BackupExecutor) fetchPreBackupPodTemplates() (*k8upv1alpha1.PreBackupPodList, error) {
+	podList := &k8upv1alpha1.PreBackupPodList{}
+
+	err := b.Client.List(b.CTX, podList, client.InNamespace(b.Obj.GetMetaObject().GetNamespace()))
+	if err != nil {
+		return nil, fmt.Errorf("could not list pod templates: %w", err)
+	}
+
+	return podList, nil
+}
+
+// generateDeployments creates a new PreBackupDeployment for each given PreBackupPod template.
+func (b *BackupExecutor) generateDeployments(templates []k8upv1alpha1.PreBackupPod) []appsv1.Deployment {
+	deployments := make([]appsv1.Deployment, 0)
+
+	for _, template := range templates {
+
+		template.Spec.Pod.PodTemplateSpec.ObjectMeta.Annotations = map[string]string{
+			cfg.Config.BackupCommandAnnotation: template.Spec.BackupCommand,
+			cfg.Config.FileExtensionAnnotation: template.Spec.FileExtension,
+		}
+
+		podLabels := map[string]string{
+			"k8up.syn.tools/backupCommandPod": "true",
+			"k8up.syn.tools/preBackupPod":     template.Name,
+		}
+
+		template.Spec.Pod.PodTemplateSpec.ObjectMeta.Labels = podLabels
+
+		deployment := appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      template.GetName(),
+				Namespace: b.Obj.GetMetaObject().GetNamespace(),
+				Labels: labels.Set{
+					"k8up.syn.tools/preBackupPod": template.Name,
+				},
+			},
+			Spec: appsv1.DeploymentSpec{
+				Replicas: pointer.Int32Ptr(1),
+				Template: template.Spec.Pod.PodTemplateSpec,
+				Selector: &metav1.LabelSelector{
+					MatchLabels: podLabels,
+				},
+			},
+		}
+
+		err := controllerutil.SetOwnerReference(b.Config.Obj.GetMetaObject(), &deployment, b.Scheme)
+		if err != nil {
+			b.Config.Log.Error(err, "cannot set the owner reference", "name", b.Config.Obj.GetMetaObject().GetName(), "namespace", b.Config.Obj.GetMetaObject().GetNamespace())
```

Comment:
Should this perhaps be either a skip or an abort condition? Because if we can't set the owner reference, this object will not be garbadge collected by k8s.

I think that skipping this particular deployment or straight-out failing this function might be the safer choice.

Reply:
tbh, I don't see a reason this error can ever occur. The docs says that the error can occur if there is another owner that has the controller flag set. If K8up creates a new deployment, there can't be another owner anyway. Since we never update a deployment (only create or delete), there shouldn't be multiple owners.
Thinking about it, I would even go as far to write `_ = controllerutil.SetOwnerReference()...`

Reply:
So I have looked into it. There's no unexpected API call, it's all local code. I'd suggest to leave it like it is then.